src/App.tsx

```diff
@@ -1,26 +1,178 @@
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<string | null>(null);
+  const [filter, setFilter] = useState(FilterStates.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const [title, setTitle] = useState('');
+
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  const fetchTodos = useCallback(async () => {
+    try {
+      const data = await getTodos();
+
+      setTodos(data);
+    } catch (err) {
+      setError(ErrorMessages.LOAD_TODOS);
+      setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+  }, []);
+
+  const handleAddTodo = useCallback(async (): Promise<void> => {
+    setIsLoading(true);
+    try {
+      const newTempTodo: Todo = {
+        id: 0,
+        userId: USER_ID,
+        title,
+        completed: false,
+      };
+
+      setTempTodo(newTempTodo);
+
+      const newTodo = await addTodo(title.trim());
+
+      setTodos(prevTodos => [...prevTodos, newTodo]);
+      setTempTodo(null);
+      setTitle('');
+    } catch {
+      setError(ErrorMessages.ADDING_TODOS);
+      setTempTodo(null);
+    } finally {
+      setIsLoading(false);
+      setTimeout(() => {
+        inputRef.current?.focus();
+      }, 100);
+    }
+  }, [title]);
+
+  const handleDeleteTodo = useCallback(
+    async (todoId: number): Promise<void> => {
+      try {
+        await deleteTodo(todoId);
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+      } catch {
+        setError(ErrorMessages.DELETING_TODOS);
+      } finally {
+        inputRef.current?.focus();
+      }
+    },
+    [],
+  );
+
+  const handleHideError = () => {
+    setError(null);
+  };
+
+  const handleClearCompleted = useCallback(async (): Promise<void> => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    try {
+      const deletionResults = await Promise.allSettled(
+        completedTodos.map(todo => handleDeleteTodo(todo.id)),
+      );
+
+      const hasErrors = deletionResults.some(
+        result => result.status === 'rejected',
+      );
+
+      if (hasErrors) {
+        setError(ErrorMessages.DELETING_SOME_TODOS);
+      }
+    } catch {
+      setError(ErrorMessages.COMPLETED_TODOS);
+    } finally {
+      inputRef.current?.focus();
+    }
+  }, [handleDeleteTodo, todos]);
+
+  const filteredTodos = useMemo(() => {
+    return todos.filter(todo => {
+      switch (filter) {
+        case FilterStates.ACTIVE:
+          return !todo.completed;
+        case FilterStates.COMPLETED:
+          return todo.completed;
+        default:
+          return true;
+      }
+    });
+  }, [todos, filter]);
+
+  const hasTodos = todos.length > 0;
+  const allCompleted = todos.length > 0 && todos.every(todo => todo.completed);
+
+  useEffect(() => {
+    if (error) {
+      setTimeout(() => {
+        setError(null);
+      }, 3000);
```

Comment:
It's a good practice to clear timeouts in the cleanup function of `useEffect `to prevent memory leaks. This ensures that any pending timeouts are cleared if the component unmounts or if the dependencies change.